torchao/csrc/cpu/paged_attention_kernel.cpp

```diff
@@ -0,0 +1,535 @@
+                         attn_mask.size(-1)})
+                  .expand({attn_mask_size_0, attn_mask_size_1, qSize, kvSize});
+}
+
+/**
+ * Performs scale-dot-product for the next token based on cached key-value
+ * attention.
+ *
+ * This function computes the attention weights and applies the attention
+ * mechanism to obtain the final output. It takes in tensors representing the
+ * query, key cache, value cache, head mapping, scale, block tables, context
+ * lengths, block size The
+ * output tensor is updated with the computed attention values.
+ *
+ * @param out           Output tensor [num_seqs, 1, num_heads, head_size].
+ * @param query         Query tensor [num_seqs, 1, num_heads, head_size].
+ * @param key_cache     The pre-allocated buffer to store the key cache. The
+ * shape should be [num_blocks, block_size, num_heads, head_size].
+ * @param value_cache   The pre-allocated buffer to store the value cache. The
+ * shape should be [num_blocks, block_size, num_heads, head_size].
+ * @param head_mapping  Head mapping tensor [num_heads]. The mapping from the
+ * query head to the kv head to support GQA/MQA. The shape should be the number
+ * of query heads.
+ * @param scale         Scaling factor for attention weights. In general, it is:
+ * float(1.0 / (head_size ** 0.5)).
+ * @param block_tables  Block tables tensor [num_seqs, max_num_blocks_per_seq].
+ * @param context_lens  Context lengths tensor [num_seqs].
+ * @param block_size    The block size which means the number of token in every
+ * block.
+ * @param max_context_len Maximum context length.
+ * @param attn_mask  Optional tensor of attention_mask
+ */
+template <typename scalar_t>
+void paged_attention_kernel(at::Tensor &out, at::Tensor &query,
+                            at::Tensor &key_cache, at::Tensor &value_cache,
+                            at::Tensor &head_mapping, const double scale,
+                            at::Tensor &block_tables, at::Tensor &context_lens,
+                            int64_t block_size,
+                            c10::optional<at::Tensor> attn_mask) {
+
+  using accum_t = at::opmath_type<scalar_t>;
+  using Vec = at::vec::Vectorized<accum_t>;
+  const auto dtype = query.scalar_type();
+  const auto accumulate_dtype = at::toOpMathType(dtype);
+  auto max_context_len = context_lens.max().item<int64_t>();
+  auto num_seqs = query.size(0);
+  auto q_len = query.size(2);
+  auto num_heads = query.size(1);
+  auto head_size = query.size(3);
+  auto num_kv_heads = key_cache.size(1);
+  auto max_num_blocks_per_seq = block_tables.size(1);
+  auto kv_head_group_size = num_heads / num_kv_heads;
+  bool has_attn_mask = attn_mask.has_value() && attn_mask.value().numel();
+  if (has_attn_mask) {
+    attn_mask.value() = attn_mask.value().to(at::kFloat);
+    reshape_attn_mask_to_4d(attn_mask.value(), num_seqs, num_heads, q_len,
+                            attn_mask.value().size(-1));
```

Comment:
Does this convert the attn_mask to 4D or just view it as 4D? Since we are working on raw pointers, perhaps we don't need to expand it as 4D view here?